perf: optimize hot paths and reduce overhead for low-end devices

negezor opened the pull request:
While profiling my website, I often see numerous `unhead` calls on hot paths on low-performance devices. Therefore, this PR aims to reduce the overall impact on performance. This PR includes:
- Compare values directly instead of allocating an array and then iterating over it https://jsperf.app/megoxe
- Use `Set` instead of an array since `set.has()` runs in constant time, unlike `array.includes()` https://jsperf.app/pikuzu/3
- Instead of using `string.startsWith()` for the first character, it's better to use index access as it is faster https://jsbench.me/welq15149d/1
- Instead of `string.split()[0]`, it's better to use `string.indexOf()` + `string.substring()` https://jsperf.app/zovuke
- Avoiding promises can greatly improve performance, as they are performance killers. I added a simple helper that allows you to not lose much in readability, but significantly improve performance. https://jsperf.app/gacuwo/3
- To the previous point, we also remove the `async` modifier from the function where a `Promise` is already returned without the need for `await`

<details>
  <summary>Benchmark Result</summary>
  <img src="https://github.com/user-attachments/assets/14e7835b-084b-4c9b-8e22-de91f012ee20">
</details>

There are a few other areas where performance can be improved, namely:
- ~Avoiding `Promise` as they hurt performance~ Implemented via the `thenable` helper. `Promise` is now seen as an edge case
- ~Reducing the number of `Object.entries()`~ Done. If we want performance, we'll have to say goodbye to `Object.entries`, `Object.values` ​​and `Object.keys`
- ~Possibly combining two sorts in the sort plugin.~ Done

Since this library is a building block for other applications, I believe it should have the maximum possible performance.


Performance was tested on the following hardware:
CPU: `AMD Ryzen 7950x3D`
System: `WSL 2 Arch Linux on Windows 11`
Node.js: `v22.5.0`

Before
```
✓ dom-useHead (1) 11339ms
  name          hz      min      max     mean      p75      p99     p995     p999     rme  samples
· x50        9.3537  99.2468  144.97   106.91   108.30   138.09   144.97   144.97  ±1.13%      100
✓ ssr bench (1) 11341ms
  name          hz      min      max     mean      p75      p99     p995     p999     rme  samples
· x50 ssr   89.0412  9.8624  18.1303  11.2308  11.4576  15.0532  15.3867  17.0052  ±0.47%     1000
```

After
```
✓ dom-useHead (1) 9408ms
  name          hz       min      max     mean      p75     p99    p995     p999      rme  samples
· x50       11.2612  80.9079   113.33  88.8004  89.7723  112.17  113.33   113.33   ±1.11%      100
✓ ssr bench (1) 7332ms
  name           hz      min      max     mean      p75     p99    p995     p999     rme   samples
· x50 ssr    138.38   6.4453  10.3630   7.2264   7.3634  9.5104  9.6495  10.1183  ±0.42%      1000
```
After using the forked hookable with these changes https://github.com/unjs/hookable/pull/102

```
✓ dom-useHead (1) 9157ms
  name           hz      min      max     mean      p75     p99    p995    p999      rme   samples
· x50       11.5724  79.0402   118.18  86.4127  87.9506  106.06  118.18  118.18   ±1.16%       100
✓ ssr bench (1) 5274ms
  name           hz      min      max     mean      p75     p99    p995    p999      rme   samples
· x50 ssr    193.42   4.6108   8.3951   5.1700   5.3634  5.9843  6.4341  8.1951   ±0.41%      1000
```

harlan-zw commented:
Woah this is really impressive work, thank you so much. I'll try get some benchmarks tomorrow but all looks good so far.

> Avoiding Promise as they hurt performance

This seems like a good idea for v2

> Since this library is a building block for other applications, I believe it should have the maximum possible performance.

Agreed!

negezor commented:
Separately, I would like to share a small victory in `processTemplateParams`. This iteration approached character-by-character parsing
Before
```
 ✓ processTemplateParams (5) 5306ms
   name                        hz     min     max    mean     p75     p99    p995    p999     rme  samples
 · basic               124,256.92  0.0068  1.2928  0.0080  0.0079  0.0188  0.0203  0.0467  ±0.71%    62129   slowest
 · nested props        223,829.82  0.0036  0.6271  0.0045  0.0042  0.0144  0.0273  0.0359  ±0.51%   111915
 · not found props     321,787.56  0.0026  0.2824  0.0031  0.0030  0.0056  0.0139  0.0278  ±0.41%   160894
 · with url          4,358,487.01  0.0002  2.3345  0.0002  0.0002  0.0003  0.0004  0.0011  ±1.03%  2179244
 · simple string    15,700,280.18  0.0000  0.4865  0.0001  0.0001  0.0001  0.0001  0.0004  ±0.71%  7850141   fastest
```
After
```
 ✓ processTemplateParams (5) 5484ms
   name                        hz     min     max    mean     p75     p99    p995    p999     rme  samples
 · basic               270,898.59  0.0029  0.1939  0.0037  0.0036  0.0048  0.0069  0.0419  ±0.30%   135450
 · nested props        261,071.41  0.0033  0.2390  0.0038  0.0038  0.0045  0.0064  0.0224  ±0.31%   130536   slowest
 · not found props     340,171.83  0.0025  1.3075  0.0029  0.0030  0.0035  0.0045  0.0148  ±0.59%   170086
 · with url          5,110,691.95  0.0002  1.8063  0.0002  0.0002  0.0002  0.0003  0.0005  ±1.09%  2555346
 · simple string    18,396,122.62  0.0000  2.2063  0.0001  0.0001  0.0001  0.0001  0.0001  ±0.91%  9198062   fastest
```

harlan-zw commented:
Just let me know when you're finished making the changes you'd like, I'll work on getting it merged and benchmarked further.

negezor commented:
@harlan-zw I have made all the changes I wanted. You can merge now 😊

harlan-zw commented:
You've done a great job on this PR, thanks again.

I've tried to do some benchmarking on these changes but the performance actually seems very slightly worst off, maybe you have a better idea on how to accurately bench this e2e though. 

**Before**
```
 ✓ test/vue/dom-useHead.bench.ts (1) 13515ms
   ✓ dom-useHead (1) 13514ms
     name      hz     min     max    mean     p75     p99    p995    p999     rme  samples
   · x50   7.9066  112.44  177.98  126.48  129.53  156.32  177.98  177.98  ±1.29%      100
 ✓ test/unhead/ssr/ssr-perf.bench.ts (1) 13381ms
   ✓ ssr bench (1) 13380ms
     name          hz      min      max     mean      p75      p99     p995     p999     rme  samples
   · x50 ssr  75.4054  12.1169  25.1118  13.2616  13.2981  17.7024  18.7785  22.4677  ±0.47%     1000
```

**After**
```
 ✓ test/vue/dom-useHead.bench.ts (1) 13672ms
   ✓ dom-useHead (1) 13671ms
     name      hz     min     max    mean     p75     p99    p995    p999     rme  samples
   · x50   7.7883  113.84  199.53  128.40  130.55  194.32  199.53  199.53  ±2.05%      100
 ✓ test/unhead/ssr/ssr-perf.bench.ts (1) 13599ms
   ✓ ssr bench (1) 13598ms
     name          hz      min      max     mean      p75      p99     p995     p999     rme  samples
   · x50 ssr  74.2828  12.2232  22.5889  13.4621  13.5607  19.1117  20.1525  22.3832  ±0.55%     1000
```

I think we can still probably merge this since we've slightly reduced the bundle size but wanted to give you the opportunity to provide benchmarks as it would be great if we could advertise a % improvement.

negezor commented:
I think it greatly depends on the specifics of the project. Typically we have more than one `useHead()` in an application, and there is also an initial `head.push()` that is contained for the application. Not to mention, we often want schema-org to be included as well. Forgot to mention that I tested with these changes https://github.com/unjs/hookable/pull/102 😅

During testing I used slowdown 6x in dev tools. For me, two factors were important in optimization:
- Initial script execution time
- Time required to process changes

A little later I will check what is slowing down in this benchmark.

negezor commented:
Through these benchmarks I was able to find other slow code sections, thanks.

On a small investigation, `Promise.all` was eating all the performance bonus
Before
![image](https://github.com/user-attachments/assets/caffe094-4839-44f2-bd40-bcaf8a30a0cd)
After
![image](https://github.com/user-attachments/assets/7a8d3d82-600b-45c6-9bd0-47a5fb644583)

Does the `tagDedupeKey` function really need a second parameter for the filter? I didn't find its use in the code.

I'll still try to improve the other slow spots

harlan-zw commented:
> Does the tagDedupeKey function really need a second parameter for the filter? I didn't find its use in the code.

I can't seem to think where this would be used so please remove it :+1:

negezor commented:
A few intermediate results of the improvements:
<details>
  <summary><b>normaliseProps</b></summary>

  Before
  <img src="https://github.com/user-attachments/assets/d53dc16b-bab3-4e71-ab81-dc9afd745887">
  After
  <img src="https://github.com/user-attachments/assets/38225947-848c-48de-8470-d29ce84215b8">
</details>
<details>
  <summary><b>normaliseEntryTags</b></summary>

  Before
  <img src="https://github.com/user-attachments/assets/ec35b745-8589-4485-8717-105a34760c23">
  After
  <img src="https://github.com/user-attachments/assets/cdedb1d9-e420-41bc-b155-aec49413b679">
  Current
  <img src="https://github.com/user-attachments/assets/f1d3d593-3cd4-41e4-afc0-cf4247cea495">
</details>

I started looking at `Promise` as an edge case, it allowed me to practically cancel out the impact of them. In version 2 of unhead I think it's clearly worth getting rid of them, because it's usually an edge case, in frameworks we have reactivity which is synchronous. And if the developer needs it, he can always do `await something()`.

Now all the emphasis is on synchronous operations and overhead from asynchronous returns from `hookable`

negezor commented:
@harlan-zw I've finished improving other areas where performance was dropping. I updated the results in the main description of the PR. Could you please review it again?

The only thing that bothers me in the `dom-useHead.bench.ts` benchmark is that all the focus is on the `jsdom` library.
<details>
  <summary>flamegraph</summary>
  <img src="https://github.com/user-attachments/assets/30e7a434-d836-4900-a2e2-b5ba31df55d7">
</details>

It turns out that if we really want more performance, we need to move away from using `Object.entries()`, `Object.values()`, and `Object.keys()` in favor of `for in` with `Object.prototype.hasOwnProperty`. This slightly worsens code readability, but it's not critical considering it gives a 2x, sometimes 3x performance boost.

Another thing consuming 1/4 of the performance is mini-loops that just checked a couple of properties to avoid code duplication. I removed them where direct checks are possible, but this doesn't add much duplication.

The overall package size has indeed increased, but executing JS is much more expensive than parsing it. We could reduce the package size by dropping support for `Promise`.

harlan-zw commented:
Hi @negezor, thanks again for your hard work on this I really appreciate it. 

I had some time to sit and make a [proper SSR benchmark](https://github.com/unjs/unhead/blob/main/test/bench/ssr-harlanzw-com-e2e.bench.ts) based on my site and all of the plugins being used. Running it on this PR it looks like you've **improved the SSR by 20% :tada: :rocket:**!

The package size difference is negligible so all good there. 

<summary>
<details>

```
-- current --
 ✓ test/bench/ssr-harlanzw-com-e2e.bench.ts (1) 3211ms
   ✓ ssr e2e bench (1) 3209ms
     name        hz     min     max    mean     p75     p99    p995    p999     rme  samples
   · e2e   1,615.77  0.5511  2.1222  0.6189  0.6177  1.0534  1.2626  1.6639  ±0.50%     5000
 ✓ test/bench/ssr-harlanzw-com-e2e.bench.ts (1) 3243ms
   ✓ ssr e2e bench (1) 3242ms
     name        hz     min     max    mean     p75     p99    p995    p999     rme  samples
   · e2e   1,598.03  0.5499  1.8620  0.6258  0.6190  1.0233  1.1068  1.3735  ±0.46%     5000
 ✓ test/bench/ssr-harlanzw-com-e2e.bench.ts (1) 3213ms
   ✓ ssr e2e bench (1) 3211ms
     name        hz     min     max    mean     p75     p99    p995    p999     rme  samples
   · e2e   1,613.98  0.5546  2.0391  0.6196  0.6200  0.9634  1.0067  1.3202  ±0.39%     5000
 ✓ test/bench/ssr-harlanzw-com-e2e.bench.ts (1) 3211ms
   ✓ ssr e2e bench (1) 3210ms
     name        hz     min     max    mean     p75     p99    p995    p999     rme  samples
   · e2e   1,614.06  0.5492  2.0267  0.6196  0.6255  1.0152  1.0993  1.5132  ±0.45%     5000
 ✓ test/bench/ssr-harlanzw-com-e2e.bench.ts (1) 3220ms
   ✓ ssr e2e bench (1) 3219ms
     name        hz     min     max    mean     p75     p99    p995    p999     rme  samples
   · e2e   1,610.08  0.5545  1.9893  0.6211  0.6159  1.0911  1.2073  1.4338  ±0.48%     5000


-- this PR  -- 

✓ test/bench/ssr-harlanzw-com-e2e.bench.ts (1) 2678ms
   ✓ ssr e2e bench (1) 2676ms
     name        hz     min     max    mean     p75     p99    p995    p999     rme  samples
   · e2e   1,949.40  0.4521  1.7332  0.5130  0.5086  0.8895  0.9698  1.2820  ±0.50%     5000


 ✓ test/bench/ssr-harlanzw-com-e2e.bench.ts (1) 2650ms
   ✓ ssr e2e bench (1) 2648ms
     name        hz     min     max    mean     p75     p99    p995    p999     rme  samples
   · e2e   1,972.05  0.4497  1.8316  0.5071  0.5046  0.8682  0.9137  1.3332  ±0.46%     5000

 ✓ test/bench/ssr-harlanzw-com-e2e.bench.ts (1) 2615ms
   ✓ ssr e2e bench (1) 2613ms
     name        hz     min     max    mean     p75     p99    p995    p999     rme  samples
   · e2e   1,998.93  0.4499  1.6762  0.5003  0.5048  0.8365  0.8900  1.0098  ±0.42%     5000

 ✓ test/bench/ssr-harlanzw-com-e2e.bench.ts (1) 2665ms
   ✓ ssr e2e bench (1) 2664ms
     name        hz     min     max    mean     p75     p99    p995    p999     rme  samples
   · e2e   1,958.39  0.4502  2.1185  0.5106  0.5057  0.9516  1.0020  1.2473  ±0.55%     5000

 ✓ test/bench/ssr-harlanzw-com-e2e.bench.ts (1) 2678ms
   ✓ ssr e2e bench (1) 2677ms
     name        hz     min     max    mean     p75     p99    p995    p999     rme  samples
   · e2e   1,950.99  0.4508  1.7268  0.5126  0.5148  0.8579  0.8979  1.0237  ±0.40%     5000

```

<details>
</summary>

As all tests are passing I'm going to merge this, I'll hold off on the release until I can do a little bit more testing, ideally, we release this as a patch but it might (?) be a little risky.

**Future Performance Opportunities**

I'd be very open to making breaking changes that will simply the API for end users, reduce internal code and improve the speed. These can be introduced in a v2 beta.

You identified removing promises would help with this, would you be interested in working on this in a separate PR? The only constraint is that the DOM rendering should still be delayed via a promise (to avoid thrashing DOM updates with incremental `useHead` calls)'.

Would be great to hear any other ideas you have also.

Btw you should setup github sponsors if you can :stuck_out_tongue:, would be great to give you a one-off amount.

negezor commented:
> Running it on this PR it looks like you've improved the SSR by 20% 🎉 🚀!

That's cool. I've also been testing with my SSR, there the numbers range from 10% for the simplest pages and up to 25% on complex ones with multiple nested routers.

> ideally, we release this as a patch but it might (?) be a little risky.

In case of any regressions, a patch can be released 😅

> You identified removing promises would help with this, would you be interested in working on this in a separate PR?

Yeah sure, now it'll just come down to deleting branches for Promise.

> Would be great to hear any other ideas you have also.

The module is great as it is, covering all my needs and covered by almost every possible test. I don't even have an idea of what to suggest 😅

> Btw you should setup github sponsors if you can 😛, would be great to give you a one-off amount.

I'm not planning to setup Github Sponsors at the moment, but thank you for the suggestion 😊

negezor commented:
I deployed the fork with these changes to production, got this result:
![image](https://github.com/user-attachments/assets/f5c98e05-2b11-4d7e-b49e-dc6edfb3a40e)
Current
![image](https://github.com/user-attachments/assets/801025a6-0f3d-4f07-a574-3333143b3cb4)
This PR
![image](https://github.com/user-attachments/assets/ba97cffa-5c5d-4d47-a776-07573a72d9b7)

harlan-zw commented:
Wow very nice! :heart_eyes: I will try get the release out next couple of days, sorry for the delay!
